nats.go

```diff
@@ -1850,6 +1849,17 @@ func (nc *Conn) doReconnect(err error) {
 
 		// Process connect logic
 		if nc.err = nc.processConnectInit(); nc.err != nil {
+			// If we have a lastErr recorded for this server
+			// do the normal processing here. We might get closed.
+			if nc.current.lastErr != nil {
+				nc.mu.Unlock()
+				nc.processErr(nc.err.Error())
```

Comment:
I don't think this is correct. Suppose that the lib reconnects to a server that does not send an -ERR, but processConnectInit() returns an error (could be a TCP error after initial connect, etc..) then processErr() will cause the connection to be closed, therefore aborting the reconnect attempts.

Let me ask you this: why do we store the lastErr in the server object in the pool? Imagine a client with 10 server urls. It is connected to 1st, then its JWT expires and first server close the connection, client will reconnect to server2, which will return JWT expiration error on connect, so it will move to server 3, etc.. Only when reaching server 1 again the logic of twice the same auth expiration error would kick in.
If we consider that the User Authentication Expired error is a function of the client JWT, not the server, then there is no reason to store the last error in `srv` struct, should be in `Conn`. We actually don't need to store `lastErr` but simply a boolean that indicates that we got that an auth error. Then another boolean that indicates that we should abort (re)connect attempts if we get an auth error and the first boolean was set. 
A simple test could be done in processConnectInit() when we know we got an -ERR and check for the 2 possible auth errors and set those flags. Here in doReconnect we would check the abort flag and simply break if set.
I can try to work on that if you prefer or we could discuss more.

The only caveat is that you treat the AUTHORIZATION_ERR and AUTHENTICATION_EXPIRED_ERR with regard to lastErr and failing after 2nd error. Unlike AUTHENTICATION_EXPIRED_ERR which I argue is a function of the client, not the server, a client could get AUTHORIZATION_ERR on one server and not on another from its pool (which would mean that the auth config is not the same in the cluster).

Reply:
I am trying to group the auths together in some fashion. The enclosing account could expire, be revoked, our user could be revoked etc.

Yes I fail when I "loop" back around and get the same error again. That is by design.

Reply:
Point is that processErr() will close the connection if given -ERR is none of the one we allow for a reconnect. Here, you may not even get a -ERR passed to processErr() since it could be that we got a error such as socket closed, etc..

Reply:
We only call it if we have a lastErr recorded.

Reply:
Please re-read at least this:
```
I don't think this is correct. Suppose that the lib reconnects to a server that does not send an -ERR, but processConnectInit() returns an error (could be a TCP error after initial connect, etc..) then processErr() will cause the connection to be closed, therefore aborting the reconnect attempts.
```

Reply:
The connection will be closed iff the connection had been established and a lastErr (which right now is auth only) was recorded.

Reply:
1- Client has S1, S2 in URLs and connects to S1 with valid JWT.
2- JWT about to expire so user updates JWT so that it is valid for next reconnect
3- JWT expires, server sends -ERR auth expired and closes conn
4- client processes async ERR and stores lastErr = AuthExpired
5- client start reconnect logic due to connection closed by server
6- client creates TCP connection to S2 but while reading from socket, server dies, socket closes. processConnectInit() returns an error (nc.err is NOT set at this point).
7- since err != nil and lastErr is set, you call processErr(nil), that will close the connection.

So client stops reconnecting while it should have go back to S1 and since JWT had been updated should succeed.

Reply:
S2 will not have a lastErr set, so it should not close and will skip that conditional and go back to S1.

Reply:
But processErr() will fall into the final `else` statement and close the connection because there is no -ERR passed to processErr()..

Reply:
actually you will even get a panic() because you call `processErr(nc.err.Error())` whil nc.err could be nil.

Reply:
S2 will not call processErr since lastErr will not be set. Maybe I am missing something and its time to jump on a GH.

Reply:
Conditional is on nc.current.lastErr, not nc.err.

Reply:
But we could have been back to S1 here where there was a recorded lastErr and still not be able to successfully return from processConnectInit() and it will close the connection and likely panic since nc.err may be cleared in this doReconnect() loop!

Reply:
nc.err is set but could get cleared after we release the lock so will snapshot.

Reply:
Apology for the ranting.. I was wrong. Yes, you set nc.err when calling nc.processConnectInit() so it won't be nil and we discussed and you are ok with connection being closed when we are back to say S1 that previously had lastErr, even if the actual nc.err is not an auth error.